src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs

```diff
@@ -294,37 +295,62 @@ private static bool IsValueTypeErrorOrBlank<T>(FormulaValue val)
             return new InMemoryTableValue(irContext, pairs.Select(pair => pair.Key));
         }
 
-        private static async IAsyncEnumerable<DValue<RecordValue>> LazyFilterAsync(
+        private static async Task<DValue<RecordValue>> LazyFilterRowAsync(
+           EvalVisitor runner,
+           SymbolContext context,
+           DValue<RecordValue> row,
+           LambdaFormulaValue filter)
+        {
+            if (row.IsValue)
+            {
+                var childContext = context.WithScopeValues(row.Value);
+
+                // Filter evals to a boolean 
+                var result = await filter.EvalAsync(runner, childContext);
+                var include = false;
+                if (result is BooleanValue booleanValue)
+                {
+                    include = booleanValue.Value;
+                }
+                else if (result is ErrorValue errorValue)
+                {
+                    return DValue<RecordValue>.Of(errorValue);
+                }
+
+                if (include)
+                {
+                    return row;
+                }
+            }
+
+            return null;
+        }
+
+        private static async Task<DValue<RecordValue>[]> LazyFilterAsync(
             EvalVisitor runner,
             SymbolContext context,
             IEnumerable<DValue<RecordValue>> sources,
-            LambdaFormulaValue filter)
+            LambdaFormulaValue filter, 
+            int topN = int.MaxValue)
```

Comment:
1. Don't see topN used anywhere in the body.
2. "topN" assumes order, filter doesn't have order spec, is the meaning here to use natural order or semantics are "takeN" (take _any_ N rows)?
3. Nit: I'd go with topN as int? to be more specific when condition is not present
4. Need to be careful with async implementation. On a large rowset we don't want to evaluate filter for all rows. We want to implement strategy when we fill N requested matches with minimal filter evals. One way is to define M as batch window size and keep hopping its evaluation until N rows are filled up